components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java

```diff
@@ -74,14 +76,101 @@ public boolean isSigned(RequestObject requestObject) {
+                    getCertificateForAlias(oAuth2Parameters.getTenantDomain(), oAuth2Parameters.getClientId());
+        } catch (RequestObjectException e) {
+            String message = "Error retrieving public certificate for service provider checking whether a jwks " +
+                    "endpoint is configured for the service provider with client_id: " + oAuth2Parameters.getClientId();
+            log.warn(message);
+            if (log.isDebugEnabled()) {
+
+                log.debug(message, e);
+            }
+        }
+        if (certificate == null) {
+            if (log.isDebugEnabled()) {
+
+                log.debug("Public certificate not configured for Service Provider with " +
+                        "client_id: " + oAuth2Parameters.getClientId() + " of tenantDomain: " + oAuth2Parameters
+                        .getTenantDomain() + ". Fetching the jwks endpoint for validating request object");
+            }
+            String jwksUri = getJWKSEndpoint(oAuth2Parameters);
+            isVerified = isSignatureVerified(jwt, jwksUri);
+        } else {
+            if (log.isDebugEnabled()) {
+
+                log.debug("Public certificate configured for Service Provider with " +
+                        "client_id: " + oAuth2Parameters.getClientId() + " of tenantDomain: " + oAuth2Parameters
+                        .getTenantDomain() + ". Using public certificate  for validating request object");
+            }
+            isVerified = isSignatureVerified(jwt, certificate);
+        }
         requestObject.setIsSignatureValid(isVerified);
         return isVerified;
     }
 
+    /**
+     * Fetch JWKS endpoint using OAuth2 Parameters.
+     *
+     * @param oAuth2Parameters  oAuth2Parameters
+     */
+    private String getJWKSEndpoint(OAuth2Parameters oAuth2Parameters) throws RequestObjectException {
+
+        String jwksUri = StringUtils.EMPTY;
+        ServiceProviderProperty[] spProperties;
+        try {
+            spProperties = OAuth2Util.getServiceProvider(oAuth2Parameters.getClientId())
+                    .getSpProperties();
+        } catch (IdentityOAuth2Exception e) {
+            throw new RequestObjectException("Error while getting the service provider for client ID " +
+                    oAuth2Parameters.getClientId(), OAuth2ErrorCodes.SERVER_ERROR, e);
+        }
+
+        if (spProperties != null) {
+            for (ServiceProviderProperty spProperty : spProperties) {
+                if (Constants.JWKS_URI.equals(spProperty.getName())) {
+                    jwksUri = spProperty.getValue();
+                    if (log.isDebugEnabled()) {
+                        log.debug("Found jwks endpoint " + jwksUri + " for service provider with client id " +
+                                oAuth2Parameters.getClientId());
+                    }
+                    break;
+                }
+            }
+        } else {
+            return StringUtils.EMPTY;
+        }
+        return jwksUri;
```

Comment:
Rather than deriving the JWKS URI using SP properties, we can get it directly using getJwksUri()[1] method of the SP.

[1] https://github.com/wso2/carbon-identity-framework/blob/1915d7dbd9a90de25229e81c1d506045e3573cfc/components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/ServiceProvider.java#L447